go/vt/vtgate/autocommit_test.go

```diff
@@ -0,0 +1,318 @@
+/*
+Copyright 2017 Google Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package vtgate
+
+import (
+	"testing"
+
+	"golang.org/x/net/context"
+
+	"github.com/youtube/vitess/go/sqltypes"
+
+	querypb "github.com/youtube/vitess/go/vt/proto/query"
+	vtgatepb "github.com/youtube/vitess/go/vt/proto/vtgate"
+)
+
+// This file contains tests for all the autocommit code paths
+// to make sure that single round-trip commits are executed
+// correctly whenever possible.
+
+// TestAutocommitUpdateSharded: insant-commit.
+func TestAutocommitUpdateSharded(t *testing.T) {
+	executor, sbc1, sbc2, _ := createExecutorEnv()
+
+	if _, err := autocommitExec(executor, "update user set a=2 where id = 1"); err != nil {
+		t.Fatal(err)
+	}
+	testBatchQuery(t, "sbc1", sbc1, &querypb.BoundQuery{
+		Sql:           "update user set a = 2 where id = 1 /* vtgate:: keyspace_id:166b40b44aba4bd6 */",
+		BindVariables: map[string]*querypb.BindVariable{},
+	})
+	testBatchQuery(t, "sbc2", sbc2, nil)
+
+	// This is the only test where we verify transaction count
+	if got, want := sbc1.AsTransactionCount.Get(), int64(1); got != want {
```

Comment:
I would personally like to see a call to `testCommitCount(t, "sbc1", sbc1, 0)` in all these cases where we expect instant-commit to hold.

It would make sure that any future changes in vtgate don't send a commit anyway even if AsTransaction was true in these cases.